credentials/tls/certprovider/pemfile/builder.go

```diff
@@ -0,0 +1,130 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package pemfile
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"path/filepath"
+	"time"
+
+	"google.golang.org/protobuf/encoding/protojson"
+	"google.golang.org/protobuf/types/known/durationpb"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+)
+
+const (
+	pluginName             = "file_watcher"
+	defaultRefreshInterval = 600 * time.Second
+)
+
+func init() {
+	certprovider.Register(&pluginBuilder{})
+}
+
+type pluginBuilder struct{}
+
+func (p *pluginBuilder) ParseConfig(c interface{}) (*certprovider.BuildableConfig, error) {
+	data, ok := c.(json.RawMessage)
+	if !ok {
+		return nil, fmt.Errorf("meshca: unsupported config type: %T", c)
+	}
+	cfg, err := pluginConfigFromJSON(data)
+	if err != nil {
+		return nil, err
+	}
+	return certprovider.NewBuildableConfig(pluginName, cfg.canonical(), func(certprovider.BuildOptions) certprovider.Provider {
+		p, err := NewProvider(Options{
+			CertFile:        cfg.certFile,
+			KeyFile:         cfg.keyFile,
+			RootFile:        cfg.keyFile,
+			RefreshDuration: cfg.refreshInterval,
+		})
+		if err != nil {
+			// This should never really happen since we validate the config in
+			// ParseConfig and we do not use the BuildOptions passed in here.
+			logger.Errorf("NewProvider(%+v) failed: %v", err)
+		}
+		return p
+	}), nil
+}
+
+func (p *pluginBuilder) Name() string {
+	return pluginName
+}
+
+type pluginConfig struct {
+	certFile        string
+	keyFile         string
+	caFile          string
+	refreshInterval time.Duration
+}
+
+func (pc *pluginConfig) canonical() []byte {
+	return []byte(fmt.Sprintf("%s:%s:%s:%s", pc.certFile, pc.keyFile, pc.caFile, pc.refreshInterval.String()))
```

Comment:
Is the explicit `.String()` necessary when `%s` is used as formatting?  (`fmt` is weird sometimes.)